server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -161,17 +161,29 @@ public void getSynonymRules(String resourceName, int from, int size, ActionListe
             .setSize(size)
             .setPreference(Preference.LOCAL.type())
             .setTrackTotalHits(true)
-            .execute(listener.delegateFailure((searchResponseListener, searchResponse) -> {
-                final long totalSynonymRules = searchResponse.getHits().getTotalHits().value;
-                if (totalSynonymRules == 0) {
-                    listener.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
-                    return;
+            .execute(new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse searchResponse) {
+                    final long totalSynonymRules = searchResponse.getHits().getTotalHits().value;
+                    if (totalSynonymRules == 0) {
+                        listener.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
+                        return;
+                    }
+                    final SynonymRule[] synonymRules = Arrays.stream(searchResponse.getHits().getHits())
+                        .map(SynonymsManagementAPIService::hitToSynonymRule)
+                        .toArray(SynonymRule[]::new);
+                    listener.onResponse(new PagedResult<>(totalSynonymRules, synonymRules));
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof IndexNotFoundException) {
+                        listener.onFailure(new ResourceNotFoundException(resourceName));
```

Comment:
Should we have the same error message as on the line 169?

Reply:
Makes sense!